File: finalyearproject/repos/dkd-platform-libs/sdk/python/src/dkd_platform/traceparent.py
# HAND-AUTHORED platform primitive (NOT dkdgen-generated).
# PL-05 — W3C traceparent parse/format/inject/extract (R6 spine correlation).
"""W3C Trace Context `traceparent` support for HTTP and the Kafka event spine.

Replaces the bare header pass-through (security.py CorrelationContext) with a real, validated
`traceparent` so a trace stitches across services and across the outbox → Kafka → inbox hop.

Wire format (W3C Trace Context level 1):

    traceparent = version "-" trace-id "-" parent-id "-" trace-flags
                =  2hex   "-"  32hex   "-"   16hex   "-"     2hex

This SDK owns only correct parse / format / inject / extract + a fresh span-id — full OTel span
creation stays a service concern. `to_header()` feeds straight into the PL-02
`OutboxRelay.headers(..., traceparent=...)` so the trace rides the event to the consumer's inbox.
"""
from __future__ import annotations

import os
import re
from dataclasses import dataclass
from typing import Iterable, Mapping, Optional

HEADER = "traceparent"

_VERSION_LEN = 2
_TRACE_ID_LEN = 32
_SPAN_ID_LEN = 16
_FLAGS_LEN = 2

_HEX = re.compile(r"\A[0-9a-f]+\Z")

_INVALID_VERSION = "ff"
_ZERO_TRACE = "0" * _TRACE_ID_LEN
_ZERO_SPAN = "0" * _SPAN_ID_LEN

FLAG_SAMPLED = 0x01


def _is_hex(s: str, length: int) -> bool:
    return len(s) == length and bool(_HEX.match(s))


@dataclass(frozen=True)
class TraceParent:
    """A parsed, validated traceparent. `trace_id`/`span_id` are lowercase hex."""

    trace_id: str
    span_id: str
    flags: int = FLAG_SAMPLED
    version: str = "00"

    @property
    def sampled(self) -> bool:
        return bool(self.flags & FLAG_SAMPLED)

    def format(self) -> str:
        """Serialize back to the wire form (round-trips with `parse`)."""
        return "%s-%s-%s-%02x" % (self.version, self.trace_id, self.span_id, self.flags & 0xFF)

    def child(self) -> "TraceParent":
        """Same trace, a fresh span-id — the correlation of a downstream call/event."""
        return TraceParent(self.trace_id, new_span_id(), self.flags, self.version)


def new_trace_id() -> str:
    return os.urandom(16).hex()


def new_span_id() -> str:
    return os.urandom(8).hex()


def new(sampled: bool = True) -> TraceParent:
    """A brand-new root traceparent (fresh trace-id + span-id)."""
    return TraceParent(new_trace_id(), new_span_id(), FLAG_SAMPLED if sampled else 0)


def parse(header: Optional[str]) -> Optional[TraceParent]:
    """Parse a traceparent header; return None on any malformed / disallowed value."""
    if not header:
        return None
    parts = header.strip().split("-")
    if len(parts) != 4:
        return None
    version, trace_id, span_id, flags = parts
    if not (_is_hex(version, _VERSION_LEN) and _is_hex(trace_id, _TRACE_ID_LEN)
            and _is_hex(span_id, _SPAN_ID_LEN) and _is_hex(flags, _FLAGS_LEN)):
        return None
    if version == _INVALID_VERSION:          # 0xff is forbidden by the spec
        return None
    if trace_id == _ZERO_TRACE or span_id == _ZERO_SPAN:  # all-zero ids are invalid
        return None
    return TraceParent(trace_id, span_id, int(flags, 16), version)


def to_header(tp: TraceParent) -> str:
    """The wire string — for an outgoing HTTP header or `OutboxRelay.headers(traceparent=...)`."""
    return tp.format()


def inject(tp: TraceParent, headers: Optional[Mapping[str, str]] = None) -> dict[str, str]:
    """Return a new header mapping with `traceparent` set (immutable: never mutates the input)."""
    out = dict(headers or {})
    out[HEADER] = tp.format()
    return out


def inject_event_headers(tp: TraceParent,
                         headers: Optional[Iterable[tuple[str, bytes]]] = None
                         ) -> list[tuple[str, bytes]]:
    """Kafka/event header list with `traceparent` appended — mirrors OutboxRelay.headers shape."""
    out = list(headers or [])
    out.append((HEADER, tp.format().encode()))
    return out


def extract(headers: Optional[Mapping[str, str]]) -> Optional[TraceParent]:
    """Inbox side: pull a traceparent out of a header mapping (case-insensitive lookup)."""
    if not headers:
        return None
    for k, v in headers.items():
        if k.lower() == HEADER:
            return parse(v)
    return None

File: finalyearproject/repos/dkd-platform-libs/sdk/csharp/Dkd.Platform/Traceparent.cs
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// PL-05: W3C Trace Context (traceparent) parse / format / inject / extract, replacing the bare
// header passthrough. Full OTel-SDK span lifecycle is a service concern; the SDK guarantees a
// correct wire header so trace context flows across HTTP calls AND event headers (wired into
// OutboxRelay.Headers). Format: "version-traceid-spanid-flags" =
//   00 - 32 hex (non-zero) - 16 hex (non-zero) - 2 hex.  Canon: R6 spine carries trace context.

namespace Dkd.Platform;

using System.Security.Cryptography;

/// <summary>A parsed W3C <c>traceparent</c>. Fields are lowercase hex of the fixed canonical widths.</summary>
public sealed record Traceparent(string Version, string TraceId, string SpanId, string Flags)
{
    /// <summary>Render back to the canonical <c>version-traceid-spanid-flags</c> wire string.</summary>
    public string Format() => $"{Version}-{TraceId}-{SpanId}-{Flags}";
}

/// <summary>
/// W3C Trace Context helpers. Parses/normalises the <c>traceparent</c> header, mints new
/// trace/span ids, and injects/extracts the header on outgoing HTTP requests and event headers.
/// </summary>
public static class W3CTrace
{
    public const string HeaderName = "traceparent";
    private const string Version00 = "00";
    private const string Sampled = "01";

    /// <summary>Parse and validate a traceparent; returns null when malformed (never throws).</summary>
    public static Traceparent? Parse(string? header)
    {
        if (string.IsNullOrEmpty(header)) return null;
        var parts = header.Split('-');
        if (parts.Length != 4) return null;

        var (ver, trace, span, flags) = (parts[0], parts[1], parts[2], parts[3]);
        if (!IsHex(ver, 2) || ver == "ff") return null;            // "ff" is the forbidden/invalid version
        if (!IsHex(trace, 32) || IsAllZero(trace)) return null;    // all-zero trace-id is invalid
        if (!IsHex(span, 16) || IsAllZero(span)) return null;      // all-zero parent-id is invalid
        if (!IsHex(flags, 2)) return null;
        return new Traceparent(ver, trace, span, flags);
    }

    /// <summary>Non-throwing parse into <paramref name="traceparent"/>; true on success.</summary>
    public static bool TryParse(string? header, out Traceparent? traceparent)
    {
        traceparent = Parse(header);
        return traceparent is not null;
    }

    /// <summary>Mint a fresh root traceparent (new 128-bit trace-id + 64-bit span-id, sampled).</summary>
    public static Traceparent NewRootSpan() => new(Version00, NewTraceId(), NewSpanId(), Sampled);

    /// <summary>A child traceparent under <paramref name="parent"/>'s trace-id with a fresh span-id.</summary>
    public static Traceparent NewChildSpan(Traceparent parent)
    {
        ArgumentNullException.ThrowIfNull(parent);
        return parent with { SpanId = NewSpanId() };
    }

    /// <summary>New 128-bit trace-id as 32 lowercase-hex chars (never all-zero).</summary>
    public static string NewTraceId() => RandomHex(16);

    /// <summary>New 64-bit span-id as 16 lowercase-hex chars (never all-zero).</summary>
    public static string NewSpanId() => RandomHex(8);

    /// <summary>Inject the traceparent into an outgoing HTTP / event header map.</summary>
    public static void Inject(IDictionary<string, string> headers, Traceparent traceparent)
    {
        ArgumentNullException.ThrowIfNull(headers);
        ArgumentNullException.ThrowIfNull(traceparent);
        headers[HeaderName] = traceparent.Format();
    }

    /// <summary>Extract and validate the traceparent from an inbound header map; null when absent/malformed.</summary>
    public static Traceparent? Extract(IReadOnlyDictionary<string, string> headers)
    {
        ArgumentNullException.ThrowIfNull(headers);
        return headers.TryGetValue(HeaderName, out var raw) ? Parse(raw) : null;
    }

    private static string RandomHex(int bytes)
    {
        Span<byte> b = stackalloc byte[16];
        b = b[..bytes];
        do { RandomNumberGenerator.Fill(b); } while (AllZero(b));
        return Convert.ToHexString(b).ToLowerInvariant();
    }

    private static bool IsHex(string s, int len)
    {
        if (s.Length != len) return false;
        foreach (var c in s)
            if (!Uri.IsHexDigit(c) || char.IsUpper(c)) return false; // canon lowercase hex only
        return true;
    }

    private static bool IsAllZero(string hex)
    {
        foreach (var c in hex) if (c != '0') return false;
        return true;
    }

    private static bool AllZero(ReadOnlySpan<byte> b)
    {
        foreach (var x in b) if (x != 0) return false;
        return true;
    }
}
